### PR TITLE
Add ClearML Logger

### PR DESCRIPTION
**Pull request by thepycoder**
Thanks a lot for all the great work you all do :)

This PR is to add a logger for [ClearML](https://github.com/allegroai/clearml).

### The PR is only tested on [this sample](https://github.com/explosion/projects/tree/v3/pipelines/tagger_parser_ud) project so far. 
I'm a spaCy newbie, so not sure if I should do more testing :)  I got the tip to clone and adapt [this project](https://github.com/explosion/projects/tree/v3/integrations/wandb) which I'll try to do if I can get the time for it.

### Metrics visualisation
Currently, I've grouped all metrics that are more than 1 layer deep on the same graph in the webUI. The reason being that giving every one of them their own separate graph, makes the UI very full of graphs. Now, there are 2 graphs that are chock-full of metrics, but they do belong near each other and it's easy to filter them within the graph itself. Personally, I think this makes sense, but I'm all for changing it like it is in the other loggers if you prefer it that way :)

Here is a gif of what the result looks like:
graphs_spacy

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
...
+`spacy.ClearMLLogger.v1` is a logger that tracks the results of each training step
+using the [ClearML](https://www.clear.ml/) tool. To use
+this logger, ClearML should be installed and you should have initialized (using the command above).
+The logger will send all the gathered information to your ClearML server, either [the hosted free tier](https://app.clear.ml) 
+or the open source [self-hosted server](https://github.com/allegroai/clearml-server). This logger captures all the following information (all visible in the webUI):
+
+- The full spaCy config file contents
+- Code information such as git repository, commit ID and uncommitted changes.
+- Full console output
+- Miscellanious info such as time, python version and hardware information
```
Comment (rmitsch) [inline README.md:166]:
```suggestion
- Miscellaneous info such as time, python version and hardware information
```

File: README.md [hunk 2]
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
...
+clearml-init
+```
+
+### Usage
+
+`spacy.ClearMLLogger.v1` is a logger that tracks the results of each training step
+using the [ClearML](https://www.clear.ml/) tool. To use
+this logger, ClearML should be installed and you should have initialized (using the command above).
+The logger will send all the gathered information to your ClearML server, either [the hosted free tier](https://app.clear.ml) 
+or the open source [self-hosted server](https://github.com/allegroai/clearml-server). This logger captures all the following information (all visible in the webUI):
```
Comment (rmitsch) [inline README.md:161]:
```suggestion
or the open source [self-hosted server](https://github.com/allegroai/clearml-server). This logger captures all the following information (all visible in the web UI):
```

File: README.md [hunk 3]
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
...
+
+- The full spaCy config file contents
+- Code information such as git repository, commit ID and uncommitted changes.
+- Full console output
+- Miscellanious info such as time, python version and hardware information
+- Output scalars:
+    - The final score is logged under the scalar `score`.
+    - Individual component scores are grouped together on 1 scalar plot, use the webUI to filter.
```
Comment (rmitsch) [inline README.md:169]:
```suggestion
    - Individual component scores are grouped together on one scalar plot, use the web UI to filter.
```

File: spacy_loggers/clearml.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    task_name: Optional[str] = None,
+    log_best_dir: Optional[str] = None,
+    log_latest_dir: Optional[str] = None,
+):
+    try:
+        # test that these are available
+        from clearml import Task, Dataset, OutputModel  # noqa: F401
+    except ImportError as exc:
+        raise ImportError(
+            "The 'mlflow' library could not be found - did you install it? "
```
Comment (rmitsch) [inline spacy_loggers/clearml.py:23]:
```suggestion
            "The 'clearml' library could not be found - did you install it? "
```

File: spacy_loggers/clearml.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                            )
+                        if log_best_dir and info["score"] == max(info["checkpoints"])[0]:
+                            best_model.update_weights_package(
+                                weights_path=log_best_dir,
+                                auto_delete_file=False,
+                                target_filename='best_model'
+                            )
+
+        def finalize() -> None:
+
```
Comment (rmitsch) [inline spacy_loggers/clearml.py:117]:
```suggestion
```

File: spacy_loggers/clearml.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
+import os
+from typing import Dict, Any, Tuple, Callable, List, Optional, IO
+import sys
+
+from spacy import util
+from spacy.training.loggers import console_logger
+
```
Comment (rmitsch) [inline spacy_loggers/clearml.py:7]:
```suggestion


```

**Review by rmitsch — CHANGES_REQUESTED** [review]
@thepycoder Thanks a lot for your contribution, looks good! The setup with ClearML works as in the readme described. Only a few nitpicks from my side. Could you be so kind and also run ` black` on your code?

Once all comments have been addressed I'll wait a few days before my approval - to see if any of the other spaCy maintainers want to review this as well.

**Review by rmitsch — APPROVED** [review]

**Comment by thepycoder** [comment]
Hi @rmitsch !

Thank you very much for the review! I just ran black on my changes and committed the result.

I also used the built-in "add suggestion to batch" functionality of github for the first time to accept your suggestions, it aparently resolved all the issues automatically, let me know if that's not what you wanted :) All comments were completely valid though, so no counterpoints on my end.

Let me know if you want me to do anything else!

File: README.md [hunk 7]
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
...
+
+### Usage
+
+`spacy.ClearMLLogger.v1` is a logger that tracks the results of each training step
+using the [ClearML](https://www.clear.ml/) tool. To use
+this logger, ClearML should be installed and you should have initialized (using the command above).
+The logger will send all the gathered information to your ClearML server, either [the hosted free tier](https://app.clear.ml) 
+or the open source [self-hosted server](https://github.com/allegroai/clearml-server). This logger captures all the following information (all visible in the web UI):
+
+- The full spaCy config file contents
```
Comment (shadeMe) [inline README.md:163]:
```suggestion
- The full spaCy config file contents.
```

File: README.md [hunk 8]
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
...
+- Code information such as git repository, commit ID and uncommitted changes.
+- Full console output
```
Comment (shadeMe) [inline README.md:165]:
```suggestion
- Full console output.
```

File: README.md [hunk 9]
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
...
+- Miscellaneous info such as time, python version and hardware information
```
Comment (shadeMe) [inline README.md:166]:
```suggestion
- Miscellaneous info such as time, python version and hardware information.
```

File: README.md [hunk 10]
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
...
+- Output scalars:
+    - The final score is logged under the scalar `score`.
+    - Individual component scores are grouped together on one scalar plot, use the web UI to filter.
```
Comment (shadeMe) [inline README.md:169]:
```suggestion
    - Individual component scores are grouped together on one scalar plot (filterable using the web UI).
```

File: README.md [hunk 11]
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
...
+    - Loss values of different components are logged with the `loss_` prefix.
+
+Then optionally, depending on the setting below, the following can be captured:
```
Comment (shadeMe) [inline README.md:172]:
```suggestion
In addition to the above, the following artifacts can also be optionally captured:
```

**Review by shadeMe — COMMENTED** [review]
Looks good overall! Some minor comments below.

File: spacy_loggers/clearml.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        nlp: "Language", stdout: IO = sys.stdout, stderr: IO = sys.stderr
+    ) -> Tuple[Callable[[Dict[str, Any]], None], Callable[[], None]]:
+        config = nlp.config.interpolate()
+        config_dot = util.dict_to_dot(config)
+        for field in remove_config_values:
+            del config_dot[field]
+        config = util.dot_to_dict(config_dot)
+        task = Task.init(
+            project_name=project_name,
+            task_name=task_name if task_name else "spaCy Training",
```
[the remark this answers is not in the document]
Comment (thepycoder) [reply spacy_loggers/clearml.py:41]: Sure!

File: spacy_loggers/clearml.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+                }
+            )
+
+        def log_step(info: Optional[Dict[str, Any]]):
+            console_log_step(info)
+            if info is not None:
+                score = info["score"]
+                other_scores = info["other_scores"]
+                losses = info["losses"]
+                task.get_logger().report_scalar(
```
[the remark this answers is not in the document]
Comment (thepycoder) [reply spacy_loggers/clearml.py:75]: You're right. I also think indexing the dict directly was bad as well. I don't quite know what can and cannot be in the info dict, but it seems safer to use `.get()` instead, so I added that change as well.

File: spacy_loggers/clearml.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (thepycoder) [reply spacy_loggers/clearml.py]: Made it non-optional in both the function argument list and updated the readme as well.

**Comment by thepycoder** [comment]
@shadeMe Thank you so much for the PR review.

First, I apologize, the code is ok, but it seems I was very sloppy with the docs in leaving e.g. MLFlow references in there. That should not have been on you guys to find!

Next, I implemented your suggestions, please check 2 of them that I left comments on.

Please Note: I also added an additional commit that changes the output dataset ID. It was an HTML link before, but it did not actually parse correctly on our end! So I'm reverting back to just the ID.

Thanks again for the work you guys put in here!

**Review by shadeMe — APPROVED** [review]

**Comment by rmitsch** [comment]
Merged :tada: Thanks again, @thepycoder! And thanks to @shadeMe for doing a more thorough review than me :pray:

**Pull request merged**